### PR TITLE
Update from facebook 1ccd0b4e67b3

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This is a 1-off manual sync from fbcode to github.
It aims at replacing #399 and additionally introduces a C++ unit test for gradient.
There are followup items in #402 to put the TC caffe2 tests in a better place for integration into fbcode.

Even after #402 there are still a few followup cleanups to address but this moves the gradient in the right direction.

### CODE REVIEW

**Comment by nicolasvasilache** [comment]
> Because we told it not to care. 

Ok, does this mean you prefer to not apply the first commit "Apply TC coding conventions" or are you fine with this commit ?

> And since line breaks in TC mark the end of statement, we really don't want to split string literals unless we have ; or some string continuation symbol in TC

[This](https://github.com/facebookresearch/TensorComprehensions/pull/401/files#diff-f0df12aba9b032b995bfa66dba1d7c49) passes for me.

File: tensor_comprehensions/pybinds/pybind_options.cc [hunk 1]
```diff
@@ -55,37 +55,49 @@ PYBIND11_MODULE(mapping_options, m) {
             }
             throw std::runtime_error("Invalid option passed");
           }),
-          "Initialize the mapping options from one of the following:\n 1. naive\n 2. pointwise\n 3. mlp\n 4. conv\n 5. group_conv\n 6. single_thread")
+          "Initialize the mapping options from one of the following:\n"
+          "1. naive\n 2. pointwise\n 3. mlp\n 4. conv\n 5. group_conv\n "
```
Comment (skimo-openhub) [inline tensor_comprehensions/pybinds/pybind_options.cc]: This drops a space in from of "1.".

Also, I would split the string after every "\n" (here and below).

**Comment by nicolasvasilache** [comment]
> 625fdb8: why add it if it is not functional?

Because this is a synchronization PR to bring the internal and github repos to the same state.
I cannot just destroy people's internal tests.
Granted there should be an OSS PR to activate those tests in CI.
cc @salexspb @mingzhe09088

File: tc/c2/convolution_op.h [hunk 2]
```diff
@@ -63,8 +63,6 @@ class TcConvolutionOp : public TcOp<T, Context, Engine> {
 
     this->tc_ = tc::makeConvolution2DTc(strideH, strideW);
     this->tcName_ = tc::CONVOLUTION2D_TC_NAME;
-    this->gradTc_ = tc::makeConvolution2DGradTc(strideH, strideW);
-    this->gradTcName_ = tc::CONVOLUTION2D_GRAD_TC_NAME;
```
Comment (skimo-openhub) [inline tc/c2/convolution_op.h]: Why is this removal part of adding support for gradient computations?
Did this code do anything useful before?
If not, it should be removed first in a separate commit.

This change also removes the single uses of `CONVOLUTION2D_GRAD_TC_NAME` and `makeConvolution2DGradTc` so they should be removed together with these lines.

File: tc/c2/tc_op.h [hunk 3]
```diff
@@ -102,12 +101,8 @@ class TcOp : public Operator<Context> {
...
       }
 
-      auto parsedTcs = tc::detail::parse(tc_);
-      CHECK_EQ(parsedTcs.count(tcName_), 1u)
-          << "attempting to access undefined function " << tcName_;
-
-      auto outTensorInfo = tc::detail::inferOutputTensorInfo(
-          parsedTcs.at(tcName_), rawInputDLTensors_);
+      auto outTensorInfo =
+        tc::inferOutputTensorInfo(tc_, tcName_, rawInputDLTensors_);
```
Comment (skimo-openhub) [inline tc/c2/tc_op.h]: independent change?

**Comment by skimo-openhub** [comment]
On Mon, May 07, 2018 at 12:49:34PM +0000, Nicolas Vasilache wrote:
> > 625fdb8: why add it if it is not functional?
> 
> Because this is a synchronization PR to bring the internal and github repos to the same state.
> I cannot just destroy people's internal tests.

I think I'm missing some context here.
Why can't the internal repo have some extra test that
is not available in the github repo?

Or is the purpose to abolish the internal repo completely
such that only one repo is left?

skimo

File: tc/c2/tc_op.h (hunk 3 above)
Comment (nicolasvasilache) [reply tc/c2/tc_op.h]: spurious change actually, thanks!

Comment (nicolasvasilache) [reply tc/c2/tc_op.h]: my bad, I misunderstood until I saw it in github, please ignore ..

File: tc/c2/tc_op.h [hunk 4]
```diff
@@ -57,10 +57,9 @@ class TcOp : public Operator<Context> {
       setupNaiveCudaMappingOptions();
     }
 
-    if (args.HasArgument("gradCudaMappingOptions")) {
-      gradCudaMappingOptions_ =
-          tc::CudaMappingOptions(args.GetSingleArgument<std::string>(
-              "gradCudaMappingOptions", "ERROR"));
+    if (args.HasArgument("gradMappingOptions")) {
+      gradCudaMappingOptions_ = tc::CudaMappingOptions(
+          args.GetSingleArgument<std::string>("gradMappingOptions", "ERROR"));
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/c2/tc_op.h]: Introduced a new commit that drops all the dead code related to gradients before introducing @mingzhe09088 's implementation.
As a consequence there is no more name change but introduction with the proper name.

File: tc/c2/convolution_op.h (hunk 2 above)
Comment (nicolasvasilache) [reply tc/c2/convolution_op.h]: addressed by introducing a commit that removes all the dead code related to inexistent gradient support. Then @mingzhe09088 's commit introduces proper C2 gradient support.
